### PR TITLE
adapter: use ECDSA keys for adapter instead of RSA

### LINKED ISSUE #69: Use different signature algorithm for SSH host key
Currently this plugin generates RSA keys:

https://github.com/hashicorp/packer-plugin-ansible/blob/97b6b771eef32c620d0ea16669a0d36aed79d025/provisioner/ansible/provisioner.go#L993

However, RSA/SHA-1 was deprecated in the latest OpenSSH release ([changelog](https://www.openssh.com/txt/release-8.8)) and my builds fail with the following error:

```
qemu.example-vm: fatal: [default]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: Unable to negotiate with 127.0.0.1 port 37115: no matching host key type
found. Their offer: ssh-rsa", "unreachable": true}
```

Please, consider changing the algorithm to something newer.

### PR DESCRIPTION
**Pull request by lbajolet-hashicorp**
When using RSA keys, for some obscure reason, the SSH connection only accepts rsa-sha1 as the signature algorithm for the key, which is now unsupported in modern OpenSSH versions.

This causes problems when using Ansible in conjunction with the proxy adapter (the default), as Ansible uses `ssh' to connect, and in turn, rejects the connection unless users manually change their config to support this, or if they add extra options to Ansible in their templates to accept these insecure methods.

To fix this, we move to ECDSA keys for the Ansible adapter. We considered using ED25519, but these keys are only supported by SSH when using the native serialisation format, and not any of the PKCS variants, making it harder to serialise as we'd need to write our own serialisation method for this format, which is brittle at best.

Closes #69

### CODE REVIEW

**Comment by lbajolet-hashicorp** [comment]
Note: tested manually using a qemu builder, we can add acceptance tests later to make sure this works in various environments.

Template reference:

```hcl2
source "qemu" "test-ssh" {
  iso_url                  = "base_images/output-debian/packer-debian" # locally built image with a plain Debian
  iso_checksum             = "none"
  disk_image               = true
  ssh_username             = "debian"
  ssh_password             = "debian"
  cpu_model                = "host"
  output_directory         = "test_output"
  headless                 = true
  cpus                     = 4
  boot_wait                = "3s"
}

build {
  sources = ["qemu.test-ssh"]

  provisioner "ansible" {
    playbook_file = "playbook.yml"
  }
}
```

**Comment by lbajolet-hashicorp** [comment]
Update: added one more option for this, `ssh_proxy_use_rsa`. This lets users fallback to using RSA keys if for some reason the new default, ECDSA, is not supported by their local SSH version.

Tested on the same configuration as mentioned above, with `ssh_proxy_use_rsa = true` in the configs, and the following as local ssh config:

```
Host 127.0.0.1
      HostKeyAlgorithms +ssh-rsa
      PubkeyAcceptedKeyTypes +ssh-rsa
```

Without the ssh config setup, I do experience the `ssh-rsa` failure to authenticate, so I would think this does work as expected.

File: provisioner/ansible/provisioner.go [hunk 1]
```diff
@@ -351,6 +363,19 @@ func (p *Provisioner) Prepare(raws ...interface{}) error {
 		p.executeAnsibleFunc = p.executeAnsible
 	}
 
+	if p.config.AdapterKeyType == "" {
+		p.config.AdapterKeyType = "ECDSA"
+	}
+	p.config.AdapterKeyType = strings.ToUpper(p.config.AdapterKeyType)
+
+	switch p.config.AdapterKeyType {
+	case "RSA", "ECDSA":
```
Comment (nywilken) [inline provisioner/ansible/provisioner.go:372]: Should we allow lower case values as well?

**Review by nywilken — APPROVED** [review]
I left one comment about allowing lower case versions of the key type. But this otherwise looks good.

File: provisioner/ansible/provisioner.go (hunk 1 above)
Comment (nywilken) [reply provisioner/ansible/provisioner.go:372]: Got it!

Comment (nywilken) [reply provisioner/ansible/provisioner.go:372]: I missed the ToUpper

**Review by nywilken — APPROVED** [review]
🚢

File: provisioner/ansible/provisioner.go (hunk 1 above)
Comment (nywilken) [reply provisioner/ansible/provisioner.go:372]: I was focusing on the docs being all upper case and I missed the ToUpper 🤦

**Comment by LKHN** [comment]
in AlmaLinux OS, We really appreciate this patch. Since `SHA1` is deprecated on AlmaLinux OS 9 and we had enable the `SHA1` back as a workaround.

Thanks!

**Pull request merged**